test/dynamo/test_backward_higher_order_ops.py

```diff
@@ -0,0 +1,250 @@
+from torch._dynamo._trace_wrapped_higher_order_op import _trace_wrapped
+from torch._dynamo.testing import normalize_gm
+from torch._dynamo.utils import counters
+from torch.fx.experimental.proxy_tensor import make_fx
+
+
+def _multiply(x):
+    return x * x
+
+
+def _graph_breaking_fn(x):
+    print("Boo!")
+    return _multiply(x)
+
+
+def _side_effect_stateful_fn2(x, obj):
+    obj.counter = obj.counter + 1
+    return _multiply(x)
+
+
+def _multiply_invoke(grad):
+    return _trace_wrapped(grad, fn=_multiply)
+
+
+def _graph_break_invoke(grad):
+    return _trace_wrapped(grad, fn=_graph_breaking_fn)
+
+
+def _side_effectful_invoke2(grad, fn):
+    return _trace_wrapped(grad, fn=fn)
+
+
+class BackwardHigherOrderOpTests(torch._dynamo.test_case.TestCase):
+    def test_invoke_in_eager(self):
+        x = torch.tensor([0.5, 0.5], requires_grad=True)
+        y = torch.tensor([0.5, 0.5], requires_grad=True)
+
+        def fn(x, y):
+            x.register_hook(_multiply_invoke)
+            return x * y
+
+        out = fn(x, y)
+        out.backward(torch.tensor([2.0, 2.0]))
+        self.assertEqual(x.grad, 2 * x)
+
+    def test_invoke_in_pt2(self):
+        for backend in ["eager", "aot_eager", "inductor"]:
+            torch._dynamo.reset()
+            x = torch.tensor([0.5, 0.5], requires_grad=True)
+            y = torch.tensor([0.5, 0.5], requires_grad=True)
+
+            def fn(x, y):
+                x.register_hook(_multiply_invoke)
+                return x * y
+
+            fn = torch._dynamo.optimize(backend)(fn)
+            out = fn(x, y)
+            out.backward(torch.tensor([2.0, 2.0]))
+            self.assertEqual(x.grad, 2 * x)
+
+    def test_invoke_make_fx_forward_contrived(self):
+        x = torch.tensor([0.5, 0.5], requires_grad=True)
+        out = make_fx(_multiply_invoke)(x)
+        self.assertEqual(out(x), torch.tensor([0.25, 0.25]))
+        actual = normalize_gm(out.print_readable(False))
+
+        expected = """\
+class _multiply_invoke(torch.nn.Module):
+    def forward(self, grad_1: f32[2]):
+        invocation: f32[2] = functools_self_invoke(grad_1);  grad_1 = None
+        assert_1: f32[2] = torch._functional_assert_tensor_metadata(invocation, (2,), (1,), torch.float32);  invocation = None
+        detach: f32[2] = torch.ops.aten.detach.default(assert_1);  assert_1 = None
+        detach_1: f32[2] = torch.ops.aten.detach.default(detach);  detach = None
+        detach_2: f32[2] = torch.ops.aten.detach.default(detach_1);  detach_1 = None
+        detach_3: f32[2] = torch.ops.aten.detach.default(detach_2);  detach_2 = None
+        return detach_3
+"""
+        self.assertExpectedInline(actual, expected)
+
+    def test_invoke_make_bw(self):
+        x = torch.tensor([0.5, 0.5], requires_grad=True)
+
+        def fwd(x):
+            z = x * x
+            return z + z
+
+        res = fwd(x)
+        res.backward(torch.tensor([1.0, 1.0]))
+        out = make_fx(_multiply_invoke)(x.grad)
```

Comment:
I don't understand. this isn't... using the backward in any interesting way right?

Reply:
Its just a make_fx test with a tensor from grad